pkg/services/store/resolver/service_test.go

```diff
@@ -0,0 +1,130 @@
+package resolver
+
+import (
+	"context"
+	"testing"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/manager/registry"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	fakeDatasources "github.com/grafana/grafana/pkg/services/datasources/fakes"
+	"github.com/grafana/grafana/pkg/services/store"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/stretchr/testify/require"
+)
+
+func TestResolver(t *testing.T) {
+	ctxOrg1 := store.ContextWithUser(context.Background(), &user.SignedInUser{OrgID: 1})
+
+	ds := &fakeDatasources.FakeDataSourceService{
+		DataSources: []*datasources.DataSource{
+			{
+				Id:        123,
+				OrgId:     1,
+				Type:      "influx",
+				Uid:       "influx-uid",
+				IsDefault: true,
+			},
+			{
+				Id:    234,
+				OrgId: 1,
+				Type:  "influx",
+				Uid:   "influx-uid2",
+				Name:  "Influx2",
+			},
+		},
+	}
+
+	p1 := &plugins.Plugin{}
+	p2 := &plugins.Plugin{}
+	p3 := &plugins.Plugin{}
+
+	p1.ID = "influx"
+	p2.ID = "heatmap"
+	p3.ID = "xyz"
+
+	pluginRegistry := registry.ProvideService()
+	_ = pluginRegistry.Add(ctxOrg1, p1)
+	_ = pluginRegistry.Add(ctxOrg1, p2)
+	_ = pluginRegistry.Add(ctxOrg1, p3)
+
+	provider := ProvideObjectReferenceResolver(ds, pluginRegistry)
+
+	scenarios := []struct {
+		name   string
+		given  *models.ObjectExternalReference
+		expect ResolutionInfo
+		err    string
+		ctx    context.Context
+	}{
+		{
+			name: "Missing datasource without type",
+			given: &models.ObjectExternalReference{
+				Kind: models.StandardKindDataSource,
+				UID:  "xyz",
+			},
+			expect: ResolutionInfo{OK: false},
+			ctx:    ctxOrg1,
+		},
+		{
+			name: "OK datasource",
+			given: &models.ObjectExternalReference{
+				Kind: models.StandardKindDataSource,
+				Type: "influx",
+				UID:  "influx-uid",
+			},
+			expect: ResolutionInfo{OK: true, Key: "influx-uid"},
+			ctx:    ctxOrg1,
+		},
+		{
+			name: "Get the default datasource",
+			given: &models.ObjectExternalReference{
+				Kind: models.StandardKindDataSource,
+			},
+			expect: ResolutionInfo{
+				OK:      true,
+				Key:     "influx-uid",
+				Warning: "type not specified",
+			},
+			ctx: ctxOrg1,
+		},
+		{
+			name: "Get the default datasource (with type)",
+			given: &models.ObjectExternalReference{
+				Kind: models.StandardKindDataSource,
+				Type: "influx",
+			},
+			expect: ResolutionInfo{
+				OK:  true,
+				Key: "influx-uid",
```

Comment:
we didn't actually resolve the given reference - should we add a warning here?

Reply:
In the frontend, having a blank value for UID just picks the default -- so I don't think anything special needs to happen.

The warning in the case above is that we really *should* indicate the type we want the default for since the saved query should only work with that type!

Reply:
right, we are picking the default rather than showing an error on the UI, but we still didn't resolve the reference. I'd treat _choosing how to treat an unresolved reference_ separately from the actual resolution logic.

Reply:
this is how the frontend says "use the default" -- it is an active choice, not a passive one (but is also the default!)